Explore Page and Filter Buttons backend and frontend

nnguyen144 opened the pull request:
*Please complete the checklist so the group can quickly understand your PR.*

### Category
- [x] **Feature**
- [ ] **Bug Fix**
- [x] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [x] **Both**

### Description
This pull request creates an explore page so users may explore posts from hubs outside of their subscriptions, as well as buttons that can be used to sort a post list by date and amount of likes, as well as filter by timestamp of the post. Both the backend and frontend is changed, with the backend changed to accompany filtering via the URL.

### Related Issue(s)
[https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub/issues/7](https://github.com/UNLV-CS472-672/2024-F-GROUP3-RebelHub/issues/7)

### New Additions
**backend\Posts\helper.py**: File that holds the helper function filter_queryset, which takes in a "self" and a queryset as parameters and uses the parameters of the URL to retrieve the post list to filter and sort the post list.
**frontend/src/app/explore/page.js**: Page holding the Explore component which displays a list of potentially all public hub posts and all private hub posts that the user is subscribed to.
**frontend/src/components/Explore/Explore.js**: Retrieves initial explore list posts and displays them. Uses PostList component from frontend\src\components\posts\post-list.tsx
**frontend/src/components/Explore/Explore.module.css**: CSS for the explore component
**frontend/src/utils/FilterButtons/FilterButtons.js**: Displays the buttons used for filtering and sorting the explore list. Can be used for other post lists as well. Holds the code used for changing the post list when the buttons are pressed. When the time range is changed, a new post list is retrieved from database. When the sort is changed, the post list is rearranged in only the frontend.
**frontend/src/utils/FilterButtons/FilterButtons.module.css**: CSS for filter buttons
**frontend/src/utils/FilterButtons/set-hot-score.js**: Helper functions that generate hot scores for all posts in the post list based off of likes, comments, and timestamp of the post.

### Modifications
**backend/Posts/models.py**: Added hot_score field in order to sort by Hot.
**backend/Posts/serializers.py**: Added hot_score to serializer.
**backend/Posts/tests.py**: Added tests for Explore and Filter Buttons.
**backend/Posts/urls.py**: Added ExploreList url to retrieve a list of posts for Explore page (public-hub posts and private-hub posts where the user is subscribed to the private hub).
**backend/Posts/views.py**: Added corresponding view for Explore, which utilizes the filter_queryset function to filter the queryset.
**backend/rebel_hubs/settings.py**: Added django-filters to INSTALLED_APPS.
**backend/requirements.txt**: Added django-filters to requirements.txt
Navbar and sidebar: Adjusted navigation bar and sidebar CSS to better fit on dynamic screen sizes. 
**frontend/src/utils/url-segments.ts**: Included a URL segment for ExploreList retrieval.

### New Dependencies
Backend: django-filters 24.3 

### Visuals
Below is a gif of the buttons in use. Note that the time range buttons do not appear to work but are in fact working as all posts were made in the past 24 hours. As such, the Hot button will produce similar results to Top.
![opera_2x7emADBiC](https://github.com/user-attachments/assets/436cbd35-a6c5-4eaa-8c29-68e4dc829d4c)


### Architecture of Your Changes
Here is a use case diagram displaying how each button interacts with the frontend and backend.
![image](https://github.com/user-attachments/assets/2f92e772-65e5-45b1-a691-4b6845c936bb)


### Usage and Integration
The filter buttons can be integrated into a hub's post list, so anyone working on hubs may want to use the filter buttons.

nnguyen144 committed:
Merge branch 'UNLV-CS472-672:master' into explore

nnguyen144 committed:
Added a new view which displays a list of posts for the explore page with filtering and ordering. Added the url for it and adjusted url-segments to include the url. Added django-filters to settings.py

nnguyen144 committed:
Added explore page.js and added initial frontend to test functionality.

nnguyen144 committed:
Adjusted views.py to include random

nnguyen144 committed:
Adjusted explore's page.js to include the sidebar and navbar. Made sidebar width into vw in order to adjust for different browser sizes.

nnguyen144 committed:
Updated explore component to use ExploreButtons in order to keep cleanliness of the code.

nnguyen144 committed:
Changed name of explore url

nnguyen144 committed:
Added tests for Explore, testing if the filters work

nnguyen144 committed:
Edited explore.js to remove unnecessary code

nnguyen144 committed:
Created helper.py file to hold the code for filtering time range and sort order. Updated views to use helper.py file's function. Set default time_range and order to none (if nothing is sent in the query parameters)

nnguyen144 committed:
Adjusted CSS of NavBar to prevent overlap with Explore list

nnguyen144 committed:
Renamed ExploreButtons to FilterButtons

nnguyen144 committed:
Fixed typo in import

nnguyen144 committed:
Fixed initial post retrieve for explore page. Renamed -timestamp to new and timestamp to old

nnguyen144 committed:
Renamed -likes to top and likes to bottom. Renamed -timestamp to new and timestamp to old. Moved annotating number_of_likes to the helper function to avoid unnecessary annotation when likes is not the sort.

nnguyen144 committed:
Added sorting by hot in backend. Added hot_score to model of Posts.

nnguyen144 committed:
Fixed typo in url-segments.ts

nnguyen144 committed:
Added hot button to explore page. Added set_hot_score function which sets the hot scores for a given post list. Awaiting edit post functionality to be merged before finishing.

nnguyen144 committed:
Added more to set_hot_score

nnguyen144 committed:
Correctly adjusted code for set-hot-score.js to use Javascript code (moved it from backend to frontend earlier but didn't switch language).

nnguyen144 committed:
Merge branch 'master' into explore

nnguyen144 committed:
Added django-filters to requirements.txt

nnguyen144 committed:
Merge branch 'explore' of https://github.com/nnguyen144/2024-F-GROUP3-RebelHub into explore

nnguyen144 committed:
Relocated FilterButtons to a new folder in utils. Made it so that FilterButtons can be used with other post lists aside from Explore.

nnguyen144 committed:
Added extra comment to url-segments.ts. Made adjustments to helper.py in order to better filter and sort posts.

nnguyen144 committed:
Merge branch 'master' into explore

nnguyen144 committed:
migration merge

nnguyen144 committed:
Relocated FilterButtons folder

thecatfour commented on frontend/src/components/FilterButtons/FilterButtons.js:
```diff
@@ -0,0 +1,108 @@
+import styles from "./FilterButtons.module.css";
+import { useState } from "react";
+import api from "../api";
```
Since you changed where FilterButtons is located, you should update the import from api.js.

barrioskevin replied:
"@/utils/api"

thecatfour commented on frontend/src/components/FilterButtons/set-hot-score.js:
```diff
@@ -0,0 +1,36 @@
+import { getCommentsListUrl } from "@/utils/url-segments";
+import api from "../api";
```
Change the import here too.

barrioskevin replied:
"@/utils/api"

thecatfour commented:
### Hot Button Issue
Clicking on the "Hot" button for the first time after making the API call to the backend doesn't change the post list for me. However, on the second and subsequent clicks, it works correctly. Here is a gif illustrating my issue:

![ezgif-2-cf5c6e214d](https://github.com/user-attachments/assets/a5bad56c-2d65-4fea-8fee-19947f3a2ece)

Ignore the "Admin Post in Hub 2" and "Post for comments" changing order in the top spot. They have the same hot score, so their order doesn't matter.

### Solution 1
One possible solution that worked for me is to call set_hot_score() in Explore.js's and FilterButtons's API call. However, this means that set_hot_score needs to be called in two different components, and we will have to remember this in future changes.

### Solution 2
Another solution that worked for me is to make changeSort() in FilterButtons an async function and do something like:
```
const placeholder = await set_hot_score(sortedPosts);
```
This means set_hot_score is only called in FilterButtons, but it makes the "Hot" button take longer to sort the posts when compared to the other buttons.

### Note
This is assuming that everyone else has this problem too. If this is only on my end, then maybe we can ignore it.

barrioskevin commented:
Hello, 
I also wanted to mention that I'm getting the same issue that Leslie described above.
The issue is kind of like how Leslie described except that sometimes the 'hot' button won't work at all.

It will show the hot list some times but not all the time.

![image](https://github.com/user-attachments/assets/05bc3281-a2df-44d6-bc10-ecb001cd74a3)

![image](https://github.com/user-attachments/assets/33516a0b-3815-4fc6-a30f-1975dd2a4105)

I think Leslie was correct in saying you need to `await` the `set_hot_score` function since you have it labeled as async. You also don't have an explicit return on your `set_hot_score` function so that could potentially be why at times the list doesn't update. 

Another thing that I wanted to point out is that you are adding a lot of duplicate logic in filtering for posts. In the backend you added in a new dependency which I believe wasn't needed since django can filter just fine. You also added in a `helper.py` file which has code that you end up not even using? since you added a `change_sort` function on client side. I suggest keeping the logic either all in the backend or all in the frontend to prevent any confusions. 

I also wanted to add while I like all the different options available on the explore page, I don't really think its good desing from a user's perspective. A lot of the buttons such as year or all time would be better suited for some sort of 'search' option. The explore page shouldn't really even need any filtering buttons because explore is the filter itself we should only be showing a certain type of posts on this page. 

I have a few questions about the hot score. You have the function that calculates the hot score on the client, It looks like you even calculate it every single time the button is pressed. Is this the case? **Are you only handling calculations and filtering on the frontend?** The logic you use to calculate this score doesn't seem all that complicated and I think would be better off used in the backend. You added a field to the posts model called hot score but since you're doing the calculations on the client **is this field ever getting set?** If the score is calculated on the backend it would make for faster retrievals of the hot score list. 

In MY opinion for this pr to be merged
* Must Have
you should clean up the backend and remove anything you no longer need
update the hot functions so the button works properly

* Nice to have
a better explore page UX in which the user doesn't have many options in filtering

* Nitpicks
the django filter doesn't look needed you can probably do without it
I would suggest most of the calculations you perform on the frontend, happen in the backend.

Thank you for your time, 
- Kevin Barrios

nnguyen144 committed:
Moved hot score calculation to backend. Removed unneeded dependency. Fixed imports.

nnguyen144 committed:
Fixed test file name issue

nnguyen144 commented:
Hi @imawful and @thecatfour, I've gone ahead and fixed the import name issue, removed unnecessary dependency, and moved the hot score calculation code from the frontend to the backend. As of my part, the code works perfectly. 

As for the buttons being potentially unnecessary for the explore page, I think you might be right. Although I will keep it as it is for now, in the future I might get rid of the time buttons specifically for the explore page.

nnguyen144 committed:
Removed unnecessary async

thecatfour commented on backend/Posts/views.py:
```diff
@@ -1,11 +1,15 @@
 from django.shortcuts import get_object_or_404
 from rest_framework.response import Response
 from rest_framework.exceptions import PermissionDenied, NotFound
+from django_filters.rest_framework import DjangoFilterBackend
```
Is this import still needed?

thecatfour commented on backend/requirements.txt:
```diff
@@ -20,3 +20,4 @@ pytest-cov==5.0.0
 pytest-django==4.8.0
 pytz==2024.2
 sqlparse==0.5.1
+django-filter==24.3
```
You mention that you removed unnecessary dependencies. Is django-filter still required, or does the hot function work without the requirement?

thecatfour commented:
### Bug
I found a small bug regarding the hot scores. Hot scores only seem to be calculated when the filter button is on "hot" and the timestamp button is changed. This means that posts can have a higher or lower than intended hot score until a user selects the hot filter and then changes the timestamp button. Here is a gif recreating the bug:

![ezgif-2-a8e0e231e4](https://github.com/user-attachments/assets/891e7d34-9532-4695-a38d-881ae6531b69)

Since users are probably going to switch between time filters with the hot button selected, I don't think this bug is necessary to fix. After a user selects the hot filter and changes the timestamp button, the posts all have their hot scores changed, so the hot filter will work for all subsequent users. 

### Conclusion
After you clarify whether the imports I pointed out are necessary, I think this is ready to be merged. It would be nice to include a test for hot scores, but we are kind of running out of time. Hot scores seem to be working correctly, so I don't think tests would be necessary.

nnguyen144 commented:
Hi @thecatfour, I've removed the unnecessary import and requirements.txt line. 
The bug that you mentioned isn't necessarily a bug. It's just that the hot score isn't updated when you switch sort buttons as hot score is updated when a new list of posts is retrieved from the backend, and switching sort buttons does not do that in order to reduce load. But switching time does retrieve the posts, as a new post list with more or less posts will need to be retrieved, which is why the hot score is properly sorted there.

nnguyen144 committed:
Fixed issues from @thecatfour

thecatfour reviewed (approved):
All my issues have been fixed, so I think we can merge this PR. The errors related to comments are fixed in the comment PRs, so as long as we merge those, then those issues will be fixed.

nnguyen144 committed:
Merge branch 'UNLV-CS472-672:master' into explore

misherwa commented:
Hey Nathan, when I opened you PR request, it mentioned that there were unapplied migrations. Is that something I am supposed to do on my end, or should this your PR already have the migrations applied?
![Screenshot 2024-11-27 124743](https://github.com/user-attachments/assets/aec0b49e-8eb5-4e2c-92d8-c297755d18aa)

I applied the migrations on me end so it works:
![Screenshot 2024-11-27 125153](https://github.com/user-attachments/assets/9b15af97-f41a-4dfa-9607-03104aac7c6c)

nnguyen144 committed:
Migrate

nnguyen144 committed:
Merge branch 'explore' of https://github.com/nnguyen144/2024-F-GROUP3-RebelHub into explore

misherwa commented:
Nice, it looks good to merge
